File: plover_vm/trace.py
"""Structured execution trace."""

from __future__ import annotations

import json
from dataclasses import asdict, dataclass, field
from typing import Any


@dataclass
class TraceEntry:
    step: int
    pc: int
    phase: int
    opcode: int
    cw: int
    regs: list[int]
    halted: bool
    extra: dict[str, Any] = field(default_factory=dict)


class Tracer:
    def __init__(self) -> None:
        self.entries: list[TraceEntry] = []
        self._step = 0

    def record(
        self,
        pc: int,
        phase: int,
        opcode: int,
        cw: int,
        regs: list[int],
        halted: bool,
        **extra: Any,
    ) -> None:
        self.entries.append(
            TraceEntry(
                step=self._step,
                pc=pc,
                phase=phase,
                opcode=opcode,
                cw=cw,
                regs=list(regs),
                halted=halted,
                extra=extra,
            )
        )
        self._step += 1

    def write_jsonl(self, path: str) -> None:
        with open(path, "w", encoding="utf-8") as f:
            for e in self.entries:
                f.write(json.dumps(asdict(e)) + "\n")

File: crates/plover_core/src/trace.rs
use std::fs::File;
use std::io::Write;
use std::path::Path;

#[derive(Clone, Debug, Default)]
pub struct TraceEntry {
    pub pc: u16,
    pub phase: u8,
    pub opcode: u8,
    pub cw: u8,
    pub regs: [u8; 4],
    pub halted: bool,
}

#[derive(Default)]
pub struct Tracer {
    entries: Vec<TraceEntry>,
}

impl Tracer {
    pub fn record(
        &mut self,
        pc: u16,
        phase: u8,
        opcode: u8,
        cw: u8,
        regs: [u8; 4],
        halted: bool,
    ) {
        self.entries.push(TraceEntry {
            pc,
            phase,
            opcode,
            cw,
            regs,
            halted,
        });
    }

    pub fn write_jsonl(&self, path: &Path) -> std::io::Result<()> {
        let mut f = File::create(path)?;
        for e in &self.entries {
            let line = serde_json::json!({
                "pc": e.pc,
                "phase": e.phase,
                "opcode": e.opcode,
                "cw": e.cw,
                "regs": e.regs,
                "halted": e.halted,
            });
            writeln!(f, "{line}")?;
        }
        Ok(())
    }
}
